### PR TITLE
fix punycode issue. Fix: https://github.com/radare/radare2/issues/4837

### PR DESCRIPTION
**Pull request by P4N74**
Fix: https://github.com/radare/radare2/issues/4837
Travis test cases  are wrong. I will fix it  and create a PR in r2-regression

### CODE REVIEW

**Comment by P4N74** [comment]
radare2-regression PR: https://github.com/radare/radare2-regressions/pull/717

File: libr/util/punycode.c [hunk 1]
```diff
@@ -10,6 +10,112 @@
...
+	int i = 0;
+	while (*(input + i)) {
+	    i++;
+	}
+	return i;
+}
+
+unsigned char *utf32toutf8 (unsigned int *input) {
+	if (!input) {
+		printf ("ERROR input is null\n");
```
Comment (radare) [inline libr/util/punycode.c]: use eprintf here

File: libr/util/punycode.c [hunk 2]
```diff
@@ -10,6 +10,112 @@
...
+		return NULL;
+	}
+
+	for (i = 0; i < len; i++) {
+		val = *(input + i);
+		if (val < 0x80) {
+			*(result + j) = (char)val;
+			j++;
+		} else if (val < 0x800) {
+			*(result + j + 1) = 0x80 | (val & 0x3f);
```
Comment (radare) [inline libr/util/punycode.c]: use result[j + 1] instead of the *() syntax

File: libr/util/punycode.c [hunk 3]
```diff
@@ -10,6 +10,112 @@
...
+			*(result + j) = 0xe0 | ((val >> 12) & 0xf);
+			j += 3;
+		} else if (val < 0x200000) {
+			*(result + j + 3) = 0x80 | (val & 0x3f);
+			*(result + j + 2) = 0x80 | ((val >> 6) & 0x3f);
+			*(result + j + 1) = 0x80 | ((val >> 12) & 0x3f);
+			*(result + j) = 0xf0 | ((val >> 18) & 0x7);
+			j += 4;
+		} else {
+			printf ("ERROR in toutf8. Seems like input is invalid\n");
```
Comment (radare) [inline libr/util/punycode.c]: eprintf here

File: libr/util/punycode.c [hunk 4]
```diff
@@ -10,6 +10,112 @@
...
+
+ut32 *utf8toutf32 (const ut8 *input) {
+	if (!input) {
+		printf ("ERROR input is null\n");
+		return NULL;
+	}
+
+	int i = 0;
+	int j = 0;
+    int inpval = 0;
```
Comment (radare) [inline libr/util/punycode.c]: mixed tabs and spaces

File: libr/util/punycode.c [hunk 5]
```diff
@@ -10,6 +10,112 @@
...
+		}
+	}
+
+	*(result + j) = 0;
+	return result;
+}
+
+ut32 *utf8toutf32 (const ut8 *input) {
+	if (!input) {
+		printf ("ERROR input is null\n");
```
Comment (radare) [inline libr/util/punycode.c]: eprintf

File: libr/util/punycode.c [hunk 6]
```diff
@@ -225,9 +344,17 @@ R_API char *r_punycode_decode(const char *src, int srclen, int *dstlen) {
...
 	}
-	*dstlen = di;
-	return dst;
+
+	finaldst = utf32toutf8 (dst);
+	free (dst);
+	if (finaldst) {
+		*dstlen = strlen ((const char *)finaldst);
+	} else {
+		printf ("ERROR: finaldst is null\n");
```
Comment (radare) [inline libr/util/punycode.c]: eprintf

File: libr/util/punycode.c [hunk 7]
```diff
@@ -225,9 +344,17 @@ R_API char *r_punycode_decode(const char *src, int srclen, int *dstlen) {
 		n += i / (di + 1);
 		i %= (di + 1);
 
-		memmove (dst+i+1, dst+i, (di-i) * sizeof(char));
+		memmove (dst+i+1, dst+i, (di-i) * sizeof(ut32));
```
Comment (radare) [inline libr/util/punycode.c]: use r2-indent or add spaces between + and -

**Comment by radare** [comment]
the pr looks good! can we have some tests? i'll merge after fixing those comments

thanks

**Comment by P4N74** [comment]
@radare tests are already been updated in the PR that I commented above : https://github.com/radare/radare2/pull/6864#issuecomment-283087675
I will fix the comments and update PR in a moment :)

**Comment by P4N74** [comment]
updated the pr and added one more test-case (the one that was mentioned in the issue) to r2-regression.

**Comment by radare** [comment]
Awesome! That was fast

**Comment by P4N74** [comment]
Atleast no more punycode issue now :smile:

**Comment by radare** [comment]
🎉

**Pull request merged**